src/utils.ts

```diff
@@ -266,6 +274,176 @@ export const getConfigFilesChangedInPR = async (pr: devOps.PullRequestProperties
   }
 };
 
+
+/**
+ * 
+ * @param pr 
+ */
+export const getChangedFilesFromPR = async (
+  pr: devOps.PullRequestProperties | undefined
+): Promise<string[]> => {
+  if (pr) {
+    const filesChanged = (await pr.diff()).map((file) => file.path);
+    return filesChanged;
+  }
+  return [];
+}; 
+
+/**
+ * get all tags who is containing the changed files
+ * @returns {array} ["tag1","tag2"]
+ * @param markDownEx 
+ * @param specsChanged 
+ */
+const getTagsForFilesChanged = (
+        markDownEx: MarkDownEx,
+        specsChanged: readonly string[]
+    ): readonly string[] => {
+        const codeBlocks = getTagsToSettingsMapping(markDownEx.markDown);
+        const tagsAffected = new Set<string>();
+
+        for (const [tag, settings] of sm.entries(codeBlocks)) {
+            // for every file in settings object, see if it matches one of the
+            // paths changed
+            const filesTouchedInTag = specsChanged.filter(
+                spec => inputFile(settings).some(inputFile => spec.includes(inputFile))
+            )
+
+            if (filesTouchedInTag.length > 0) {
+                tagsAffected.add(tag)
+            }
+        }
+        return [...tagsAffected]
+    }
+
+
+/**
+ * Gets the path to the readme starting with "specification" folder and without returning the file name
+ * @param readmeUrl Full Url path to readme
+ */
+const getReadMeRelativePathToRepoWithoutFileName = (readmeUrl: string): string => {
+  return readmeUrl.substring(readmeUrl.indexOf("specification"), readmeUrl.indexOf("readme.md"));
+}
+
+export const isTagExisting = (config:string,tag:string):boolean => {
+    const content = fs.readFileSync(config, { encoding: "utf8" });
+    const readme = parse(content);
+   return getInputFilesForTag(readme.markDown,tag) != undefined
+} 
+
+/**
+ * @return {map} ,key is the readme and value is an array of the changed files belonging to the readme  
+ * [["1/readme.md",["file1.json","file2.json"]]] 
+ * @param filesChanged
+ */
+export const getChangeFilesReadmeMap = async (
+         filesChanged: string[]
+       ): Promise<Map<string, string[]>> => {
+         const configFiles = new Map<string, string[]>();
+         for (let fileChanged of filesChanged) {
+           const backup = fileChanged;
+           while (fileChanged.startsWith("specification")) {
+             if (fileChanged.toLowerCase().endsWith("readme.md") && fs.existsSync(fileChanged)) {
+               if (configFiles.has(fileChanged)) {
+               } 
+               else {
+                 configFiles.set(fileChanged, []);
+               }
+               if (!backup.endsWith("readme.md")) {
+                 let value = configFiles.get(fileChanged);
+                 if (value) {
+                   value.push(backup);
+                 }
+               }
+               break;
+             }
+             // select parent readme
+             const parts = fileChanged.split("/");
```

Comment:
Pls refactor code here. Nodejs standard lib may solve it.

Reply:
There is a little different with the avocado, I don't want to re-implement this function